File: cs/BoyerAlgorithm/BoyerAlgorithm/Pattern.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace BoyerAlgorithm
{
    class Pattern
    {
        public int[] Delta1 { get; private set; }
        public int[] Delta2 { get; private set; }

        public string Str { get; private set; }

        private void fillDelta1()
        {
            for (int i = 0; i < 256; i++)
            {
                Delta1[i] = Str.Length;
            }

            for (int i = Str.Length - 1; i >= 0; i--)
            {
                if (Delta1[Convert.ToInt32(Str[i])] == Str.Length)
                {
                    Delta1[Convert.ToInt32(Str[i])] = i;
                }
            }

            for (int i = 0; i < 256; i++)
            {
                if (Delta1[i] != Str.Length)
                {
                    Delta1[i] = Str.Length - 1 - Delta1[i];
                }
            }
        }

        private int rpr(int i)
        {
            int subPatternLength = Str.Length - i - 1;

            int j = i - 1;
            for (; j >= -subPatternLength; j--)
            {
                if (j >= 0)
                {
                    if (string.Compare(Str, j, Str, i + 1, subPatternLength) == 0)
                    {
                        if (j == 0 || Str[j - 1] != Str[i])
                        {
                            break;
                        }
                    }
                }
                else
                {
                    if (string.Compare(Str, 0, Str, i + 1 - j, subPatternLength + j) == 0)
                    {
                        break;
                    }
                }
            }

            return j;
        }

        private void fillDelta2()
        {
            Delta2[Str.Length - 1] = 1;
            for (int i = Str.Length - 2; i >= 0; i--)
            {
                Delta2[i] = Str.Length - rpr(i);
            }
        }

        public Pattern(string str)
        {
            Str = str;

            Delta1 = new int[256];
            Delta2 = new int[Str.Length];
        }

        public void PreProcess()
        {
            fillDelta1();
            fillDelta2();
        }
    }
}

File: cpp/BoyerAlgorithm/BoyerAlgorithm/pattern.cpp
#include "pattern.h"

void Pattern::fillDelta1() {
    for (int i = 0; i < 256; i++) {
        delta1[i] = (int) str.length();
    }

    for (int i = (int) str.length() - 1; i >= 0; i--) {
        if (delta1[(int) str[i]] == (int) str.length()) {
            delta1[(int) str[i]] = i;
        }
    }

    for (int i = 0; i < 256; i++) {
        if (delta1[i] != (int) str.length()) {
            delta1[i] = (int) str.length() - 1 - delta1[i];
        }
    }
}

int Pattern::rpr(int i) {
	int subPatternLength = (int) str.length() - i - 1;
    int j = i - 1;
    for (; j >= -subPatternLength; j--) {
        if (j >= 0) {
            if (str.compare(j, subPatternLength, str, i + 1, subPatternLength) == 0) {
                if (j == 0 || str[j - 1] != str[i]) {
                    break;
                }
            }
        } else {
			if (str.compare(0, subPatternLength + j, str, i + 1 - j, subPatternLength + j) == 0) {
                break;
            }
        }
    }
    return j;
}

void Pattern::fillDelta2() {
    delta2[(int) str.length() - 1] = 1;
    for (int i = (int) str.length() - 2; i >= 0; i--) {
        delta2[i] = (int) str.length() - rpr(i);
    }
}

void Pattern::preprocess() {
    fillDelta1();
    fillDelta2();
}
